src/statsdproto.rs

```diff
@@ -0,0 +1,385 @@
+use bytes::BufMut;
+use bytes::Bytes;
+use memchr::{memchr, memchr2};
+use thiserror::Error;
+
+use std::{
+    convert::{TryFrom, TryInto},
+    vec,
+};
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum Type {
+    Counter,
+    Timer,
+    Gauge,
+    DirectGauge,
+    Set,
+}
+
+impl TryFrom<&[u8]> for Type {
+    type Error = ParseError;
+
+    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
+        if value.len() < 1 && value.len() > 2 {
+            return Err(ParseError::InvalidType);
+        }
+        match value[0] {
+            b'c' => Ok(Type::Counter),
+            b'm' => Ok(Type::Timer),
+            b'g' => Ok(Type::Gauge),
+            b'G' => Ok(Type::DirectGauge),
+            b's' => Ok(Type::Set),
+            _ => Err(ParseError::InvalidType),
+        }
+    }
+}
+
+#[derive(Error, Debug)]
+pub enum ParseError {
+    #[error("invalid parsed value")]
+    InvalidValue,
+    #[error("invalid sample rate")]
+    InvalidSampleRate,
+    #[error("invalid type")]
+    InvalidType,
+    #[error("invalid tag")]
+    InvalidTag,
+}
+
+/// A Tag is a key:value pair of metric tags
+#[derive(Debug, Clone, PartialEq, PartialOrd)]
+pub struct Tag {
+    name: Vec<u8>,
+    value: Vec<u8>,
+}
+
+/// Parsed gives an owned structure which represents a parsed statsd protocol
+/// unit which owns all of its fields. When parsing, no canonicalization is
+/// performed by default.
+#[derive(Debug, Clone)]
+pub struct Parsed {
+    name: Vec<u8>,
+    mtype: Type,
+    value: f64,
+    sample_rate: Option<f64>,
+    tags: Vec<Tag>,
+}
+
+fn parse_tags(input: &[u8]) -> Result<Vec<Tag>, ParseError> {
+    // Its impossible to have a tag of less than "a:b", also short circuit for
+    // no tags
+    match input.len() {
+        len if len == 0 => return Ok(vec![]),
+        len if len < 3 => return Err(ParseError::InvalidTag),
+        _ => (),
+    };
+
+    let mut tags: Vec<Tag> = Vec::new();
+    let mut scan = input;
+    loop {
```

Comment:
Ordinarily I'd suggest a parser combinator for this kind of thing, but since you're after performance first (per the crate docs on statsdproto)... C-style makes sense to me!

Reply:
I tested with `nom` originally, and its not bad at all, but felt harder to develop at the time. Slices are pretty powerful as a base abstraction so you don't lose your mind.